### ISSUE TITLE
builtins.hashFile discards string context

### ISSUE THREAD

**Issue by exFalso** [NixOS/nix#4015]
**Describe the bug**

`builtins.hashFile` discards the string context, making `builtings.hashFile ${something}` fail.

**Steps To Reproduce**

```bash
$ nix-instantiate -E 'let pkgs = import <nixpkgs> {}; in pkgs.writeText "hello.sha256" (builtins.hashFile "sha256" (pkgs.writeText "hello" "world"))'
error: opening file '/nix/store/x4vggq06hhyxpzy168lhqyn2sy5h8j2q-hello': No such file or directory
```

**Expected behavior**

No failure

**`nix-env --version` output**

```
$ nix-env --version
nix-env (Nix) 2.3.6
```

**Additional context**

I'm fairly certain the culprit is https://github.com/NixOS/nix/blob/2a8017092025a7c108ca1a829a8b4f0bcf7ee2c0/src/libexpr/primops.cc#L1363, which seems to have been lifted verbatim from https://github.com/NixOS/nix/blob/2a8017092025a7c108ca1a829a8b4f0bcf7ee2c0/src/libexpr/primops.cc#L3038

**Comment by Ericson2314** [comment]
Yeah this will need to block on the context being realized, or if we don't want to do that just fail.

**Comment by kvtb** [comment]
Actually, no.

`builtins.hashFile` does not discard the string context (although I would expect it to discard and I found this issue a moment before filine my own about "builtins.hashFile DOES NOT discard string context"):

```
nix-repl> builtins.hasContext (builtins.hashFile "sha256" (builtins.fetchurl "https://github.com/NixOS/nixpkgs/archive/release-21.05.tar.gz")) 
true

nix-repl> builtins.hasContext (builtins.hashString "sha256" (builtins.fetchurl "https://github.com/NixOS/nixpkgs/archive/release-21.05.tar.gz"))
true
```

but string contexts have nothing to do with your problem.

you just want your
 `pkgs.writeText "hello.sha256" (builtins.hashFile "sha256" (pkgs.writeText "hello" "world"))`
to trigger IFD like
 `pkgs.writeText "hello.sha256" (builtins.readFile (pkgs.writeText "hello" "world"))`
does.

**Comment by exFalso** [comment]
> Actually, no.
> 
> `builtins.hashFile` does not discard the string context (although I would expect it to discard and I found this issue a moment before filing my own issue about "builtins.hashFile DOES NOT discard string context"):
> 
> ```
> nix-repl> builtins.hasContext (builtins.hashFile "sha256" (builtins.fetchurl "https://github.com/NixOS/nixpkgs/archive/release-21.05.tar.gz")) 
> true
> 
> nix-repl> builtins.hasContext (builtins.hashString "sha256" (builtins.fetchurl "https://github.com/NixOS/nixpkgs/archive/release-21.05.tar.gz"))
> true
> ```
> 
> but string contexts have nothing to do with your problem.
> 
> you just want your
> `pkgs.writeText "hello.sha256" (builtins.hashFile "sha256" (pkgs.writeText "hello" "world"))`
> to trigger IFD like
> `pkgs.writeText "hello.sha256" (builtins.readFile (pkgs.writeText "hello" "world"))`
> does.

Hmm how is this an IFD? I thought IFD is when you import a nix expression that is itself a derivation result, no?

**Comment by kvtb** [comment]
> Hmm how is this an IFD? I thought IFD is when you import a nix expression that is itself a derivation result, no?

`builtins.readFile (derivation { .... })` is IFD too. It is easier to see it as a basic block of IFD 

`import (derivation { .... })`  == `eval (builtins.readFile (derivation { .... }))` 

`builtins.hashFile algo (derivation { .... })`  == `builtins.hashString algo (builtins.readFile (derivation { .... }))`

**Comment by kvtb** [comment]
And `pkgs.writeText` has `derivation` deep inside

**Comment by exFalso** [comment]
Hmm I still don't see how this is an IFD. The important bit of IFDs is `eval`, or in other words the primitive that makes nix a monadic language (instead of applicative), and consequently what causes the "staged" evaluation, and also why it's banned from nixpkgs.
Whereas `builtins.readFile (derivation {..})` does not do this, it creates a straight "static" dependency chain. There are no "unknown" dependency links that you can only get to after realizing the derivation.

**Comment by kvtb** [comment]
IFD is about running build and then use its result in Nix code.
you can run Nix with maxima verbosity (`-vvvvvv` command line switch), it will print when doing IFD

**Comment by exFalso** [comment]
Ah you're right, it's easy to see, as you can e.g. branch on the the file contents, which means you need to realize before continuing the evaluation

**Comment by exFalso** [comment]
There's even a section on this https://nixos.wiki/wiki/Import_From_Derivation

**Comment by edolstra** [comment]
Fixed in c6fa7775de413a799b9a137dceced5dcf0f5e6ed.
